src/function3d/smooth_periodic_function.hpp

```diff
@@ -209,24 +209,25 @@ class Smooth_periodic_function
 
         auto frg_ptr = (spfft_->local_slice_size() == 0) ? nullptr : &f_rg_[0];
 
+        using precision_type = typename std::conditional<std::is_same<T, double>::value, double, float>::type;
         switch (direction__) {
             case 1: {
                 if (gvecp_->comm_ortho_fft().size() != 1) {
                     gather_f_pw_fft();
                 }
-                spfft_->backward(reinterpret_cast<double const*>(f_pw_fft_.at(sddk::memory_t::host)), SPFFT_PU_HOST);
+                spfft_->backward(reinterpret_cast<precision_type const*>(f_pw_fft_.at(sddk::memory_t::host)), SPFFT_PU_HOST);
```

Comment:
The transform handle type of `spfft_` will also have to be changed. `spfft::TransformFloat` instead of `spfft::Transform` for single precision. This will have to be changed in several places, like Simulation_context.

Reply:
Good spot! I propose that we address the FFT interoperability in the next PR. This PR is a preparation for testing  inner() product in FP32. With FFTs we'll probably have to create second instance with FP32 because we will likely store two sets of wave-functions - double and float